simulator/bypass/data_bypass.h

```diff
@@ -90,75 +87,93 @@ class DataBypass
                 std::array<RegisterInfo, REG_NUM_MAX> array = {};
         };
 
+        // returns current stage of passed register
+        // in accordance with the current state of the scoreboard
         auto get_current_stage( RegNum num) const
         {
             return scoreboard.get_entry( num).current_stage;
         }
 
-        void update_register_info( const MIPSInstr& instr, RegNum num,
-                                   const Cycle& cycle,
-                                   RegisterStage new_dst_stage);
+        // introduces a source register of a passed instruction to scoreboard 
+        void trace_new_register( const MIPSInstr& instr, RegNum num);
 
-        void set_initial_state_to_reg_info( RegNum num)
+        // discards the information about passed register
+        void untrace_register( RegNum num)
         {
             auto& entry = scoreboard.get_entry( num);
+
             entry.current_stage = RegisterStage::in_RF();
-            entry.is_bypassible = false; 
+            entry.is_bypassible = false;
+            
+            traced_registers.erase( num); 
         }
 
         Scoreboard scoreboard = {};
+        std::set<RegNum> traced_registers = {};
 
-        /* gives the information whether bypassed data should be transformed
-           when bypassing is needed for HI register
-        */ 
+        // gives an idea whether bypassed data should be transformed
+        // when bypassing is needed for HI register 
         bool is_HI_master_DIVMULT = false;
```

Comment:
I advise you to think about how to remove that field.

Reply:
The reason: if our pipeline was very deep, you may have several HI_LO producers and consumers in-flight.